### PR TITLE
AWS signing as REST authentication option

### PR DESCRIPTION
**Pull request by nhw76**
Support AWS signing as REST authentication option

### CODE REVIEW

**Comment by nhw76** [comment]
Opening a PR as requested [here](https://github.com/open-policy-agent/opa/issues/1340)

I started to take a look at adopting the plugins code after a bit of a break; I ran straight into a dependency cycle between the plugins and rest packages; I will take more of a look later but it doesn't look like it's trivial to break.

File: plugins/rest/rest.go [hunk 1]
```diff
@@ -157,43 +100,16 @@ func New(config []byte, opts ...func(*Client)) (Client, error) {
 		return Client{}, err
 	}
 
-	tlsConfig, err := parsedConfig.validateAndInjectDefaults()
+	parsedConfig.URL = strings.TrimRight(parsedConfig.URL, "/")
```
Comment (patrick-east) [inline plugins/rest/rest.go]: Why do we have to trim the URL? Is this something that should always be done or plugin specific?

**Comment by patrick-east** [comment]
Just noticed the linter is complaining about the doc strings on public methods too https://travis-ci.org/open-policy-agent/opa/builds/532429106?utm_source=github_status&utm_medium=notification

File: plugins/rest/rest.go (hunk 1 above)
Comment (nhw76) [reply plugins/rest/rest.go]: I wasn't sure exactly what that was meant for, but I didn't want to change that as I was worried about introducing a regression. It used to be in validateAndInjectDefaults: 
https://github.com/open-policy-agent/opa/blob/3263f54a74242f2225d0d7520f401805970c86db/plugins/rest/rest.go#L51

**Comment by nhw76** [comment]
> Just noticed the linter is complaining about the doc strings on public methods too https://travis-ci.org/open-policy-agent/opa/builds/532429106?utm_source=github_status&utm_medium=notification

Apologies for that; this was expected to be an intermediate state during the refactoring.

I can either document them all or rename them so they're no longer exported - either way is fine for me.

**Comment by nhw76** [comment]
Just checking in here:

- Please let me know if you have any preference on clearing up that lint noise
- Documentation and test helper items were addressed; let me know if any further follow-ups.

**Comment by tsandall** [comment]
@nhw76 @patrick-east is OOO this week so his response will be a bit delayed. I'd like to have him give the final OK.

Regarding the linter messages, I'd recommend unexporting if that will work otherwise, yes, please document each one. I recommend unexporting so that we can keep the package API surface area as small as possible.

Thanks!

**Comment by rupertbg** [comment]
Was _just_ cloning the repo to add this, glad I checked the existing PRs! Let me know if there's anything I can do to help get this over the line. @nhw76 I see the metadata service is hardcoded here: https://github.com/open-policy-agent/opa/pull/1425/files#diff-a97fbb67bc81bdda6a7da1e0d8cf5ccbR25

Would it be worth allowing for use of the ECS Task Metadata endpoint as well? Looks like the latest version is set via an Environment Variable (https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint-v3.html)

**Comment by nhw76** [comment]
@tsandall thanks; I've eliminated the lint noise by unexporting reasonably aggressively. If that needs to change during a later refactor, then it's not a complex change.

**Comment by nhw76** [comment]
> Would it be worth allowing for use of the ECS Task Metadata endpoint as well? Looks like the latest version is set via an Environment Variable (https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint-v3.html)

@rupertbg I pushed some commits onto the branch to support that; I haven't had a change to test thoroughly; would be interested to get feedback.

File: plugins/rest/rest.go (hunk 1 above)
Comment (patrick-east) [reply plugins/rest/rest.go]: Oh interesting, I missed that we were already doing it before 😄 

It is kinda weird but I do agree we shouldn't change the behavior.

**Comment by nhw76** [comment]
Squashed.

**Pull request merged**